### PR TITLE
Cassandra ChatMessageHistory and Cassandra VectorStore enhancements

### PR DESCRIPTION
**Pull request by mieslep**
- Adds: CassandraChatMessageHistory
- Adds: Vector types of dot_product and euclildean to the default cosine
- Adds: Filter searches support operators other than just = 
- Adds: Support for composite primary keys, with partition definition
- Adds: Support for WITH clause on CREATE TABLE
- Fixes: Similarity score is returned, instead of just ranking
- Fixes: Performance enhancements (prepared statements, removing unnecessary vector value return)

### CODE REVIEW

File: langchain/src/vectorstores/cassandra.ts [hunk 1]
```diff
@@ -215,54 +280,172 @@ export class CassandraStore extends VectorStore {
...
+                          ${
+                            this.metadataColumns.length > 0
+                              ? ", " +
+                                this.metadataColumns
+                                  .map((col) => `${col.name}`)
+                                  .join(", ")
+                              : ""
+                          }`;
+
+    cql = `CREATE CUSTOM INDEX IF NOT EXISTS idx_vector_${this.table}
```
Comment (jacoblee93) [inline langchain/src/vectorstores/cassandra.ts:313]: Any injection risk?

File: langchain/src/stores/message/cassandra.ts [hunk 2]
```diff
@@ -0,0 +1,129 @@
+import { Client, DseClientOptions } from "cassandra-driver";
```
Comment (jacoblee93) [inline langchain/src/stores/message/cassandra.ts:1]: You'll need to add an entrypoint in `create-entrypoints.js` for this:

https://github.com/langchain-ai/langchainjs/blob/main/.github/contributing/INTEGRATIONS.md

**Review by jacoblee93 — CHANGES_REQUESTED** [review]
All seems reasonable, two small comments

File: langchain/src/stores/message/cassandra.ts (hunk 2 above)
Comment (mieslep) [reply langchain/src/stores/message/cassandra.ts:1]: done

File: langchain/src/vectorstores/cassandra.ts (hunk 1 above)
Comment (mieslep) [reply langchain/src/vectorstores/cassandra.ts:313]: none added - this is part of the vector store "create" process. The PR does resolve injection risks though, as previously the values were being used as part of the literal string, and the update uses bind variables instead.

File: langchain/src/vectorstores/tests/cassandra.int.test.ts [hunk 3]
```diff
@@ -1,34 +1,49 @@
 /* eslint-disable no-process-env */
```
[the remark this answers is not in the document]
Comment (mieslep) [reply langchain/src/vectorstores/tests/cassandra.int.test.ts:1]: these are in .skip .int tests

**Comment by mieslep** [comment]
@jacoblee93 questions addressed, still haven't fully worked out how to rebase on top of main, as whenever I try I end up pulling in hundreds of unrelated files to the PR :/

**Comment by jacoblee93** [comment]
All good, thank you!

**Review by jacoblee93 — APPROVED** [review]

**Pull request merged**